src/Compilers/Core/Portable/PEWriter/ExtendedPEBuilder.cs

```diff
@@ -0,0 +1,100 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Diagnostics;
+using System.Reflection.Metadata;
+using System.Reflection.Metadata.Ecma335;
+using System.Reflection.PortableExecutable;
+using Microsoft.CodeAnalysis;
+
+namespace Microsoft.Cci
+{
+    /// <summary>
+    /// This PEBuilder adds an .mvid section.
+    /// </summary>
+    internal sealed class ExtendedPEBuilder
+        : ManagedPEBuilder
+    {
+        private const string MvidSectionName = ".mvid";
+        public const int SizeOfGuid = 16;
+
+        // When the section is built with a placeholder, the placeholder blob is saved for later fixing up.
+        private Blob _mvidSectionFixup = default(Blob);
+
+        // Only include the .mvid section in ref assemblies
+        private bool _withMvidSection;
+
+        public ExtendedPEBuilder(
+            PEHeaderBuilder header,
+            MetadataRootBuilder metadataRootBuilder,
+            BlobBuilder ilStream,
+            BlobBuilder mappedFieldData,
+            BlobBuilder managedResources,
+            ResourceSectionBuilder nativeResources,
+            DebugDirectoryBuilder debugDirectoryBuilder,
+            int strongNameSignatureSize,
+            MethodDefinitionHandle entryPoint,
+            CorFlags flags,
+            Func<IEnumerable<Blob>, BlobContentId> deterministicIdProvider,
+            bool withMvidSection)
+            : base(header, metadataRootBuilder, ilStream, mappedFieldData, managedResources, nativeResources,
+                  debugDirectoryBuilder, strongNameSignatureSize, entryPoint, flags, deterministicIdProvider)
+        {
+            _withMvidSection = withMvidSection;
+        }
+
+        protected override ImmutableArray<Section> CreateSections()
+        {
+            var baseSections = base.CreateSections();
+
+            if (_withMvidSection)
+            {
+                var builder = ArrayBuilder<Section>.GetInstance(baseSections.Length + 1);
+
+                builder.Add(new Section(MvidSectionName, SectionCharacteristics.MemRead |
+                    SectionCharacteristics.ContainsInitializedData |
+                    SectionCharacteristics.MemDiscardable));
```

Comment:
@jcouv This might be handy for testing such scenarios: https://source.dot.net/#System.Reflection.Metadata.Tests/PortableExecutable/PEBuilderTests.cs,482